### PR TITLE
Fixes #31833 (Issues should have canonical meta tag)

### PR DESCRIPTION
**Pull request by chrisgraham**

### CODE REVIEW

File: core/html_api.php [hunk 1]
```diff
@@ -311,6 +311,22 @@ function html_meta_redirect( $p_url, $p_time = null, $p_sanitize = true ) {
 	return true;
 }
 
+/**
+ * Print a canonical meta tag.
+ *
+ * @param string  $p_url      The canonical URL: has to be a relative path.
+ * @return boolean
```
Comment (dregad) [inline core/html_api.php:318]: Why set a return type ? It is not being used.

File: core/layout_api.php [hunk 2]
```diff
@@ -47,16 +47,20 @@
 
 /**
  * Print the page header section
- * @param string $p_page_title   Html page title.
- * @param string $p_redirect_url URL to redirect to if necessary.
- * @param string $p_page_id      The page id.
+ * @param string $p_page_title    Html page title.
+ * @param string $p_redirect_url  URL to redirect to if necessary.
+ * @param string $p_page_id       The page id.
+ * @param string $p_canonical_url Canonical URL if necessary.
```
Comment (dregad) [inline core/layout_api.php:53]: Probably worth repeating that a relative URL is expected here (same as PHPDoc of new html_meta_canonical() function).

File: core/html_api.php [hunk 3]
```diff
@@ -311,6 +311,22 @@ function html_meta_redirect( $p_url, $p_time = null, $p_sanitize = true ) {
 	return true;
 }
 
+/**
+ * Print a canonical meta tag.
+ *
+ * @param string  $p_url      The canonical URL: has to be a relative path.
```
Comment (dregad) [inline core/html_api.php:317]: Suggest referencing to what the URL must be relative to: 
`The canonical URL: has to be relative to {@see $g_path}`

File: core/html_api.php [hunk 4]
```diff
@@ -311,6 +311,22 @@ function html_meta_redirect( $p_url, $p_time = null, $p_sanitize = true ) {
...
+ * @return boolean
+ */
+function html_meta_canonical( $p_url ) {
+	$t_url = config_get_global( 'path' ) . $p_url;
+
+	$t_url = htmlspecialchars( $t_url );
+
+	echo "\t" . '<link rel="canonical" href="' . $t_url . '" />' . "\n";
+
+	return true;
```
Comment (dregad) [inline core/html_api.php:327]: Return not needed

**Review by dregad — CHANGES_REQUESTED** [review]

File: core/html_api.php (hunk 1 above)
Comment (chrisgraham) [reply core/html_api.php]: Was just copying `html_meta_redirect`, but in hindsight that had a reason.

File: core/layout_api.php (hunk 2 above)
Comment (chrisgraham) [reply core/layout_api.php]: Ok. Made change for `$p_redirect_url` as well for consistency.

**Comment by chrisgraham** [comment]
Ok I've made those changes.

File: core/html_api.php [hunk 5]
```diff
@@ -283,7 +283,7 @@ function html_css_cdn_link( $p_url, $p_hash = '' ) {
  * $p_time is the number of seconds to wait before redirecting.
  * If we have handled any errors on this page return false and don't redirect.
  *
- * @param string  $p_url      The page to redirect: has to be a relative path.
+ * @param string  $p_url      The page to redirect: has to be relative to the install path (@see $g_path).
```
Comment (dregad) [inline core/html_api.php:286]: It really needs to be `{@see $g_path}`, with curly braces (see [PHPDoc documentation](https://docs.phpdoc.org/3.0/guide/guides/docblocks.html#description))

File: core/html_api.php [hunk 6]
```diff
@@ -311,6 +311,19 @@ function html_meta_redirect( $p_url, $p_time = null, $p_sanitize = true ) {
 	return true;
 }
 
+/**
+ * Print a canonical meta tag.
+ *
+ * @param string  $p_url      The canonical URL: has to be relative to the install path (@see $g_path).
```
Comment (dregad) [inline core/html_api.php:317]: See above comment

File: core/layout_api.php [hunk 7]
```diff
@@ -47,16 +47,20 @@
 
 /**
  * Print the page header section
- * @param string $p_page_title   Html page title.
- * @param string $p_redirect_url URL to redirect to if necessary.
- * @param string $p_page_id      The page id.
+ * @param string $p_page_title    Html page title.
+ * @param string $p_redirect_url  URL to redirect to if necessary: has to be relative to the install path (@see $g_path).
+ * @param string $p_page_id       The page id.
+ * @param string $p_canonical_url Canonical URL if necessary: has to be relative to the install path (@see $g_path).
```
Comment (dregad) [inline core/layout_api.php:53]: And again

**Review by dregad — CHANGES_REQUESTED** [review]

**Comment by chrisgraham** [comment]
Ok, thanks. Done.

**Review by dregad — APPROVED** [review]
Thanks for your contribution, I'll merge that shortly.

**Pull request merged**